client/lib/date/index.js

```diff
@@ -300,18 +300,139 @@ export const getDateDifferenceInDays = ( date, date2 ) => {
  * Get the previous date for either the previous period of year.
  *
  * @param {String} date - Base date
- * @param {Int} difference - The difference in days for the previous period. See `getDateDifferenceInDays`.
+ * @param {String} date1 - primary start
+ * @param {String} date2 - secondary start
  * @param {String} compare - `previous_period`  or `previous_year`
+ * @param {String} interval - interval
  * @return {String}  - Calculated date
  */
-export const getPreviousDate = ( date, difference, compare ) => {
+export const getPreviousDate = ( date, date1, date2, compare, interval ) => {
 	const dateMoment = toMoment( isoDateFormat, formatDate( 'Y-m-d', date ) );
+	const _date1 = toMoment( isoDateFormat, formatDate( 'Y-m-d', date1 ) );
+	const _date2 = toMoment( isoDateFormat, formatDate( 'Y-m-d', date2 ) );
 	if ( 'previous_period' === compare ) {
-		return dateMoment.clone().subtract( difference, 'days' );
+		const difference = _date1.diff( _date2, interval );
+		return dateMoment.clone().subtract( difference, interval );
 	}
 	return dateMoment.clone().subtract( 1, 'years' );
 };
 
+/**
+ * Returns the allowed selectable intervals for  a specific query.
+ *
+ * @param  {Object} query Current query
+ * @return {Array} Array containing allowed intervals.
+ */
+export function getAllowedIntervalsForQuery( query ) {
+	let allowed = [];
+	if ( 'custom' === query.period ) {
+		const { primary } = getCurrentDates( query );
+		const differenceInDays = getDateDifferenceInDays( primary.before, primary.after );
+		if ( differenceInDays > 728 ) {
+			allowed = [ 'month', 'quarter', 'year' ];
```

Comment:
Since there is some duplication of interval arrays, should we just have a list of `const`s at the top of the file that we can use to assign to `allowed` depending on the case?

```
const INTERVALS_OVER_TWO_YEARS = [ 'month', 'quarter', 'year' ];
const INTERVALS_DAY_ONLY = [ 'day' ];
...
```

Reply:
I started to go down this route, but now I think the benefit of readability trumps pulling out these arrays to const, especially if they are at the top of the page. Lets revisit this if another function will need to make use of these values?